### PR TITLE
fix: resolve pandas 3 DatetimeIndex-to-int64 unit regression

### PR DESCRIPTION
**Pull request by ebrattli**
## Problem

In pandas 3, `DatetimeIndex.to_numpy(np.int64)`, `.astype(np.int64)`, and `np.array(index, dtype=np.int64)` return the index's **native resolution** (often microseconds) rather than nanoseconds. Code that assumed nanoseconds silently produced wrong values. `Timestamp.value` is also deprecated and will eventually be removed.

Affected functions:
- `data_quality.extreme` — wrong x-axis seconds, broken hat matrix on rank-deficient input
- `data_quality.gaps_identification_{z_scores,modified_z_scores,iqr}` — implicit unit in timestamp diffs
- `detect.oscillation_detector` — `Timestamp.value` list comprehension (deprecated)
- `statistics.outliers_v1` — `.astype(np.int64)` on DatetimeIndex before standardization

## Fix

Add `datetime_index_to_ns(index)` in `ts_utils/utility_functions.py` as a single explicit conversion point (`index.as_unit("ns").asi8`), and use it everywhere. Also replace `np.linalg.inv` with `np.linalg.pinv` in the hat matrix calculation to handle rank-deficient inputs without raising `LinAlgError`.

All fixes restore pre-pandas-3 behavior (which was implicitly nanoseconds). The `pinv` change is the only intentional behavioral difference — it's strictly more robust for edge cases.

## Tests added

- Direct unit test for `datetime_index_to_ns` verifying nanosecond output
- Gap detection test asserting correct **absolute** gap duration (would have caught the unit bug)
- Oscillation detector test with a 2024 timestamp index verifying correct frequencies (would catch `Timestamp.value` removal)
- `extreme()` regression tests for modern timestamps and rank-deficient input

### CODE REVIEW

File: indsl/data_quality/outliers.py [hunk 1]
```diff
@@ -291,6 +292,5 @@ def _calculate_hat_diagonal(x: np.ndarray) -> np.ndarray:
 
     """
     X_mat = np.vstack((np.ones_like(x), x)).T
-    X_hat = X_mat @ np.linalg.inv(X_mat.T @ X_mat) @ X_mat.T
-
+    X_hat = X_mat @ np.linalg.pinv(X_mat)
     return X_hat.diagonal()
```
[the remark this answers is not in the document]
Comment (ebrattli) [reply indsl/data_quality/outliers.py]: Fixed in https://github.com/cognitedata/indsl/commit/b472575

Replaced `X_mat @ np.linalg.pinv(X_mat)` with `np.sum(X_mat * pinv_X.T, axis=1)` — same O(n·p) pinv computation, O(n) memory instead of O(n²), no intermediate n×n matrix.

File: indsl/ts_utils/utility_functions.py [hunk 2]
```diff
@@ -231,9 +240,9 @@ def get_timestamps(series: pd.Series, unit: TimeUnits = "ms") -> pd.Series:
         UserTypeError: unit is not a string
     """
     if unit == "ns":
-        values = series.index.to_numpy("datetime64[ns]").view(np.int64)
+        values = datetime_index_to_ns(series.index)
     else:
-        values = series.index.as_unit("ns").asi8 / pd.Timedelta(1, unit=unit).value
+        values = datetime_index_to_ns(series.index) / pd.Timedelta(1, unit=unit).value
```
[the remark this answers is not in the document]
Comment (ebrattli) [reply indsl/ts_utils/utility_functions.py]: Fixed in https://github.com/cognitedata/indsl/commit/b472575

Used `pd.Timedelta(1, unit=unit) // pd.Timedelta(1, "ns")` rather than the suggested `.as_unit("ns").asi8` — the latter is an `Index` property and does not exist on a scalar `Timedelta`. Integer division gives the same nanosecond count exactly without floating-point risk.

**Review by arwassa — DISMISSED** [review]

**Review by arwassa — DISMISSED** [review]

File: indsl/data_quality/outliers.py [hunk 3]
```diff
@@ -251,8 +251,9 @@ def _split_timeseries_into_time_and_value_arrays(data: pd.Series) -> tuple:
             x -> Datetime index converted to integers starting at 0
             y -> Time series values
     """
-    x = (np.array(data.index, dtype=np.int64) - data.index[0].value) / 1e9
-    y = data.to_numpy()  # Just to please pandas devs
+    index_ns = data.index.as_unit("ns").astype(np.int64).to_numpy()
```
Comment (nithinb) [inline indsl/data_quality/outliers.py]: also why not use datetime_index_to_ns?

**Review by nithinb — CHANGES_REQUESTED** [review]
In [outliers.py](https://github.com/cognitedata/indsl/blob/66af27b0395aa844838929b8c99f93e7eb577962/indsl/statistics/outliers.py#L269-L270) this issue continues to exist. This can be addressed as part of this PR

File: indsl/data_quality/outliers.py (hunk 3 above)
Comment (ebrattli) [reply indsl/data_quality/outliers.py]: Fixed in https://github.com/cognitedata/indsl/commit/298d3c4

Replaced the inline `.as_unit("ns").astype(np.int64).to_numpy()` with `datetime_index_to_ns` — same semantics, consistent with all other call sites in this PR.

**Comment by ebrattli** [comment]
Fixed in https://github.com/cognitedata/indsl/commit/298d3c4

`indsl/statistics/outliers.py:269` — replaced `.to_series().astype(np.int64)` with `datetime_index_to_ns`, same fix as was applied to `outliers_v1.py` in the original commit. Also took the opportunity to apply `datetime_index_to_ns` consistently in `data_quality/outliers.py` as suggested in the inline thread.

**Review by arwassa — DISMISSED** [review]

File: tests/data_quality/test_studentised_residuals.py [hunk 4]
```diff
@@ -10,6 +10,48 @@
...
+    data = pd.Series(dup_sig, index=dup_index)
+    # Should not raise LinAlgError; any UserValueError from validation is acceptable
+    try:
+        result = extreme(data)
+        assert isinstance(result, pd.Series)
+    except Exception as exc:
+        from indsl.exceptions import UserValueError
+
+        assert isinstance(exc, UserValueError), f"Expected UserValueError, got {type(exc)}: {exc}"
+
```
Comment (nithinb) [inline tests/data_quality/test_studentised_residuals.py]: nit and not blocking: This can be replaced to 


```suggestion
    with pytest.raises(UserValueError):
        extreme(data)

```

**Review by nithinb — COMMENTED** [review]
:unicorn: 

LGTM
had a nit but that is not blocking.

File: tests/data_quality/test_studentised_residuals.py (hunk 4 above)
Comment (ebrattli) [reply tests/data_quality/test_studentised_residuals.py]: Fixed in https://github.com/cognitedata/indsl/commit/beda183a8cd2e5c74b82e9559177ad8b1235711f

Replaced the try/except with a direct assertion. Since `pinv` handles rank-deficient input gracefully, `extreme()` succeeds on duplicate timestamps — no exception path to branch on.

Comment (nithinb) [reply tests/data_quality/test_studentised_residuals.py]: ahh, nice

**Comment by nithinb** [comment]
risk review ok

**Review by mhaider110 — APPROVED** [review]

**Review by nithinb — APPROVED** [review]
🦄
